ext/json/json.c

```diff
@@ -207,14 +212,50 @@ PHP_JSON_API int php_json_encode(smart_str *buf, zval *val, int options) /* {{{
 }
 /* }}} */
 
+static const char *php_json_get_error_msg(php_json_error_code error_code) /* {{{ */
+{
+	switch(error_code) {
+		case PHP_JSON_ERROR_NONE:
+			return "No error";
+		case PHP_JSON_ERROR_DEPTH:
+			return "Maximum stack depth exceeded";
+		case PHP_JSON_ERROR_STATE_MISMATCH:
+			return "State mismatch (invalid or malformed JSON)";
+		case PHP_JSON_ERROR_CTRL_CHAR:
+			return "Control character error, possibly incorrectly encoded";
+		case PHP_JSON_ERROR_SYNTAX:
+			return "Syntax error";
+		case PHP_JSON_ERROR_UTF8:
+			return "Malformed UTF-8 characters, possibly incorrectly encoded";
+		case PHP_JSON_ERROR_RECURSION:
+			return "Recursion detected";
+		case PHP_JSON_ERROR_INF_OR_NAN:
+			return "Inf and NaN cannot be JSON encoded";
+		case PHP_JSON_ERROR_UNSUPPORTED_TYPE:
+			return "Type is not supported";
+		case PHP_JSON_ERROR_INVALID_PROPERTY_NAME:
+			return "The decoded property name is invalid";
+		case PHP_JSON_ERROR_UTF16:
+			return "Single unpaired UTF-16 surrogate in unicode escape";
+		default:
+			return "Unknown error";
+	}
+}
+/* }}} */
+
 PHP_JSON_API int php_json_decode_ex(zval *return_value, char *str, size_t str_len, zend_long options, zend_long depth) /* {{{ */
 {
 	php_json_parser parser;
 
 	php_json_parser_init(&parser, return_value, str, str_len, (int)options, (int)depth);
 
 	if (php_json_yyparse(&parser)) {
-		JSON_G(error_code) = php_json_parser_error_code(&parser);
+		php_json_error_code error_code = php_json_parser_error_code(&parser);
+		if (0 == (options & PHP_JSON_THROW_EXCEPTIONS)) {
+			JSON_G(error_code) = error_code;
```

Comment:
Well, you can get the code and message from the exception. Why dirty the global error state in that case?